File: packages/provide-uterm/src/provide/uterm/control/plane/memory/transaction.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, TypeVar

if TYPE_CHECKING:
    import asyncio
    from collections.abc import MutableMapping

    from provide.uterm.control.plane.approval.types import ApprovalRecord
    from provide.uterm.control.plane.lease.types import LeaseRecord
    from provide.uterm.control.plane.session.types import SessionRecord
    from provide.uterm.control.plane.token.types import ResumeTokenRecord, SessionTokenRecord

_K = TypeVar("_K")
_V = TypeVar("_V")


@dataclass(slots=True)
class MemoryState:
    session_tokens: dict[tuple[str, str], SessionTokenRecord] = field(default_factory=dict)
    resume_tokens: dict[str, ResumeTokenRecord] = field(default_factory=dict)
    sessions: dict[str, SessionRecord] = field(default_factory=dict)
    approvals: dict[str, ApprovalRecord] = field(default_factory=dict)
    leases: dict[str, LeaseRecord] = field(default_factory=dict)


def _copy_state(state: MemoryState) -> MemoryState:
    return MemoryState(
        session_tokens=state.session_tokens.copy(),
        resume_tokens=state.resume_tokens.copy(),
        sessions=state.sessions.copy(),
        approvals=state.approvals.copy(),
        leases=state.leases.copy(),
    )


def _merge_table(
    root: MutableMapping[_K, _V],
    snapshot: MutableMapping[_K, _V],
    working: MutableMapping[_K, _V],
) -> None:
    """Apply only this transaction's key-level changes onto the shared table."""
    for key in set(snapshot) | set(working):
        before = snapshot.get(key)
        after = working.get(key)
        if after == before:
            continue
        if key in working:
            root[key] = working[key]
        else:
            root.pop(key, None)


@dataclass(slots=True)
class MemoryTransaction:
    _root: MemoryState
    _lock: asyncio.Lock
    state: MemoryState = field(init=False)
    closed: bool = False
    _snapshot: MemoryState = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self._snapshot = _copy_state(self._root)
        self.state = _copy_state(self._root)

    async def commit(self) -> None:
        if self.closed:
            return
        async with self._lock:
            _merge_table(self._root.session_tokens, self._snapshot.session_tokens, self.state.session_tokens)
            _merge_table(self._root.resume_tokens, self._snapshot.resume_tokens, self.state.resume_tokens)
            _merge_table(self._root.sessions, self._snapshot.sessions, self.state.sessions)
            _merge_table(self._root.approvals, self._snapshot.approvals, self.state.approvals)
            _merge_table(self._root.leases, self._snapshot.leases, self.state.leases)
        self.closed = True

    async def rollback(self) -> None:
        self.closed = True

File: packages/provide-uterm-go/controlplane/memory/transaction.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package memory

import (
	"context"
	"sync"

	cp "github.com/provide-io/provide-uterm/packages/provide-uterm-go/controlplane"
)

// Transaction is a snapshot-isolated in-memory transaction. Port of
// control.plane.memory.transaction.MemoryTransaction.
//
// On begin it captures a snapshot of the root tables and an independent working
// copy (State). Stores mutate the working copy. Commit takes the shared lock,
// runs optimistic-concurrency conflict detection across all tables, and — only
// if conflict-free — merges this transaction's key-level changes onto root.
type Transaction struct {
	root     *State
	lock     *sync.Mutex
	snapshot *State
	state    *State
	closed   bool
}

// newTransaction builds a Transaction snapshotting root under lock. The caller
// (engine.Begin) already holds nothing; Python's MemoryTransaction snapshots
// eagerly in __post_init__ without the lock, matching this constructor.
func newTransaction(root *State, lock *sync.Mutex) *Transaction {
	return &Transaction{
		root:     root,
		lock:     lock,
		snapshot: root.copyTables(),
		state:    root.copyTables(),
	}
}

// Commit persists this transaction's changes, or fails with a conflict error if
// a key it wrote was changed by a concurrently committed transaction. Idempotent
// after close.
func (t *Transaction) Commit(_ context.Context) error {
	if t.closed {
		return nil
	}
	t.lock.Lock()
	defer t.lock.Unlock()

	// Detect across all tables FIRST so a conflict aborts before any partial
	// merge is applied.
	conflict := detectConflict(t.root.SessionTokens, t.snapshot.SessionTokens, t.state.SessionTokens) ||
		detectConflict(t.root.ResumeTokens, t.snapshot.ResumeTokens, t.state.ResumeTokens) ||
		detectConflict(t.root.Sessions, t.snapshot.Sessions, t.state.Sessions) ||
		detectConflict(t.root.Approvals, t.snapshot.Approvals, t.state.Approvals) ||
		detectConflict(t.root.Leases, t.snapshot.Leases, t.state.Leases) ||
		detectConflict(t.root.GraphicalTargets, t.snapshot.GraphicalTargets, t.state.GraphicalTargets)
	if conflict {
		t.closed = true
		return cp.ConflictError("memory control-plane transaction conflicts with a concurrent commit")
	}
	mergeTable(t.root.SessionTokens, t.snapshot.SessionTokens, t.state.SessionTokens)
	mergeTable(t.root.ResumeTokens, t.snapshot.ResumeTokens, t.state.ResumeTokens)
	mergeTable(t.root.Sessions, t.snapshot.Sessions, t.state.Sessions)
	mergeTable(t.root.Approvals, t.snapshot.Approvals, t.state.Approvals)
	mergeTable(t.root.Leases, t.snapshot.Leases, t.state.Leases)
	mergeTable(t.root.GraphicalTargets, t.snapshot.GraphicalTargets, t.state.GraphicalTargets)
	t.closed = true
	return nil
}

// Rollback discards this transaction's working copy.
func (t *Transaction) Rollback(_ context.Context) error {
	t.closed = true
	return nil
}

// State exposes this transaction's working copy. It exists so tests can mirror
// the Python tests that reach into “tx.state“ (e.g. to drop a key and assert
// the delete branch of the merge). Production code goes through the stores.
func (t *Transaction) State() *State { return t.state }
